dev_tools/change-log.py

```diff
@@ -0,0 +1,332 @@
+
+
+def categorise(fragments: Iterable[Fragment], field: str) -> Dict[str, List[Fragment]]:
+    """Categorise fragments by a shared field"""
+    key = operator.attrgetter(field)
+    fragments = sorted(fragments, key=key)
+    return {
+        category: list(group)
+        for category, group in itertools.groupby(
+            fragments, key=key
+        )
+    }
+
+
+def load_metadata(fragment_dir: str) -> Tuple[List[Release], Dict[str, List[Fragment]]]:
+    r"""
+    Load all metadata from ``fragment_dir``
+
+    Returns a list of currently stored :py:class:`Release`\ s
+    and a mapping from version number to :py:class:`Fragment`\ s.
+    """
+    releases = Release.from_file(os.path.join(fragment_dir, 'versions.yaml'))
+    versioned_fragments = categorise(
+        (
+            Fragment.from_file(path)
+            for path in glob.glob(os.path.join(fragment_dir, '*.yaml'))
+            if os.path.basename(path) != 'versions.yaml'
+        ),
+        'version',
+    )
+    if UNRELEASED.semver in versioned_fragments:
+        releases.append(UNRELEASED)
+    releases.sort()
+    return releases, versioned_fragments
+
+
+def underline(line: str, symbol: str) -> List[str]:
+    """Underline a single-line ``.rst`` string"""
+    return [line, symbol * len(line), '']
+
+
+# Release information update
+def release_changes(fragment_dir, new_ver: str):
+    releases, versioned_fragments = load_metadata(fragment_dir=fragment_dir)
+    # no releases, no fragments
+    if not releases:
+        Release.to_file(
+            os.path.join(fragment_dir, 'versions.yaml'),
+            [Release(new_ver, TODAY)]
+        )
+    else:
+        if releases[0] == UNRELEASED:
+            releases[0] = Release(new_ver, TODAY)
+            for fragment in versioned_fragments[UNRELEASED.semver]:
+                fragment._replace(version=new_ver).to_file()
+        else:
+            releases.insert(0, Release(new_ver, TODAY))
+        Release.to_file(
+            os.path.join(fragment_dir, 'versions.yaml'),
+            releases
+        )
+
+
+# Changelog compilation
+def format_release(
+        release: Release,
+        fragments: List[Fragment],
+        item_format: str,
+        categories: List[str],
+) -> List[str]:
+    """Compile the changelog section for a single release"""
+    lines = underline(f'[{release.semver}] - {release.date}', '=')
+    categorised_fragments = {
+        cat.casefold(): frags
+        for cat, frags in categorise(fragments, 'category').items()
+    }
+    for category in categories:
+        caseless = category.casefold()
+        if caseless not in categorised_fragments:
+            continue
+        lines.extend(underline(category, '-'))
+        for fragment in categorised_fragments[caseless]:
+            lines.append(item_format.format(**fragment._asdict()))
+        lines.append('')
+    return lines
+
+
+CHANGELOG_HEADER = f"""
+.. Created by {APP_NAME} at {TODAY}, command
```

Comment:
It looks a bit strange that command is standing on the first line and not together with the command.